src/recovery/logical.c

```diff
@@ -513,38 +517,56 @@ orioledb_decode(LogicalDecodingContext *ctx, XLogRecordBuffer *buf)
 
 	while (ptr < endPtr)
 	{
+		const char *rec_type_str;
 		XLogRecPtr	changeXLogPtr = startXLogPtr + (ptr - startPtr);
 
 		rec_type = *ptr;
 		ptr++;
 
-		elog(DEBUG4, "RECEIVE: %s", rec_type == WAL_REC_XID ? "XID" :
-			 rec_type == WAL_REC_COMMIT ? "COMMIT" :
-			 rec_type == WAL_REC_ROLLBACK ? "ROLLBACK" :
-			 rec_type == WAL_REC_JOINT_COMMIT ? "JOINT COMMIT" :
-			 rec_type == WAL_REC_RELATION ? "RELATION" :
-			 rec_type == WAL_REC_O_TABLES_META_LOCK ? "META LOCK" :
-			 rec_type == WAL_REC_O_TABLES_META_UNLOCK ? "META_UNLOCK" :
-			 rec_type == WAL_REC_TRUNCATE ? "TRUNCATE" :
-			 rec_type == WAL_REC_SAVEPOINT ? "SAVEPOINT" :
-			 rec_type == WAL_REC_ROLLBACK_TO_SAVEPOINT ? "ROLLBACK TO SAVEPOINT" :
-			 rec_type == WAL_REC_INSERT ? "INSERT" :
-			 rec_type == WAL_REC_UPDATE ? "UPDATE" :
-			 rec_type == WAL_REC_DELETE ? "DELETE" :
-			 rec_type == WAL_REC_REINSERT ? "REINSERT" : "_UNKNOWN");
+		rec_type_str = wal_record_type_to_string(rec_type);
+
+		elog(DEBUG4, "RECEIVE record type %d (%s)", rec_type, rec_type_str);
 
 		if (rec_type == WAL_REC_XID)
 		{
-			CSNSnapshotData *csnSnapshot;
+			ptr = wal_parse_rec_xid(ptr, &oxid, &logicalXid, &heapXid);
 
-			memcpy(&oxid, ptr, sizeof(oxid));
-			ptr += sizeof(oxid);
+			if (TransactionIdIsValid(logicalXid))
+			{
+				CSNSnapshotData *csnSnapshot;
 
-			memcpy(&logicalXid, ptr, sizeof(TransactionId));
-			ptr += sizeof(TransactionId);
+				elog(DEBUG4, "RECEIVE record type %d (%s) oxid %lu logicalXId %u heapXid %u", rec_type, rec_type_str, oxid, logicalXid, heapXid);
 
-			csnSnapshot = SnapBuildGetCSNSnaphot(ctx->snapshot_builder);
-			csnSnapshot->nextXid = Max(csnSnapshot->nextXid, oxid);
+				csnSnapshot = SnapBuildGetCSNSnaphot(ctx->snapshot_builder);
+				csnSnapshot->nextXid = Max(csnSnapshot->nextXid, oxid);
+			}
+			else
+			{
+				elog(DEBUG4, "IGNORED record type %d (%s) invalid logicalXid for oxid %lu heapXid %u", rec_type, rec_type_str, oxid, heapXid);
+			}
+		}
+		else if (rec_type == WAL_REC_SWITCH_LOGICAL_XID)
+		{
+			TransactionId oldXid,
+						newXid;
+
+			ptr = wal_parse_rec_switch_logical_xid(ptr, &oldXid, &newXid);
+
+			elog(DEBUG4, "RECEIVE record type %d (%s) %u=>%u oxid %lu logicalXId %u heapXid %u",
+				 rec_type, rec_type_str,
+				 oldXid, newXid,
+				 oxid, logicalXid, heapXid);
+
+			Assert(TransactionIdIsValid(oldXid));
+			Assert(TransactionIdIsValid(newXid));
+			Assert(logicalXid == oldXid);
+			Assert(heapXid == newXid);
```

Comment:
I don't think this makes sense.  WAL_REC_SWITCH_LOGICAL_XID is especially for the case we didn't know heapXid in advance.